### ISSUE TITLE
validateCells() performance issue

### ISSUE THREAD

**Issue by budnix** [handsontable/handsontable#3276]
Validating table with 100 rows and 100 columns takes to much time. On my computer its take ~2s. See [demo](http://jsfiddle.net/ya34cgay/1/).

**Comment by krzysztofspilka** [comment]
On my old ASUS it takes "I stole 0.632 seconds of your computer time!". Do you have maybe a rough idea how we could improve that?

**Comment by budnix** [comment]
I've no idea at the moment :)

**Comment by jwhitehorn** [comment]
I too am interested in any performance improvements you can find. I have a table with neatly 1,000 rows, and several dozen columns, and from profiling it cell rendering and validation takes a lot of time. More than I'd prefer to see at least, especially when cells need to be frequently updated.

Specifically the numericValidator appears to be the worst, at least in my case.

**Comment by gstanislav** [comment]
This probably happens because of async rerender of each validated cell:

https://github.com/handsontable/handsontable/blob/master/src/core.js#L1084

**Comment by rayokao** [comment]
validateCells seems to contain plenty of iteration and function calls. Validating the entire table gets heavy especially in larger tables or when complex validators are used. However, there seems to be an undocumented function validateCell that can be used to validate single cells. I used this one to only validate the cells of the row that was modified. Seemed to help a lot.

**Comment by Whythawye** [comment]
I am validating Very Large files (10k to 100k lines), and trying to run benchmarks on how long it is taking. So far 10,000 rows took 0:12:13 and 20,000 rows took 2:10:57. When I tried 100k rows it didn't even get to the validation before chrome started complaining about a hung script. What kind of progress is being done to make this more efficient? In the readme of the trello board for the 2017 roadmap the vision for handsontable is stated as creating a "spreadsheet framework which works without limitation in performance, regardless of the size of the data set." (https://trello.com/c/pdNVL2si/71-readme) So I presume it is some sort of priority to be handle data sets of this magnitude. This is mission-critical for a large project I'm working on.

**Comment by AMBudnik** [comment]
Hi @Whythawye 

if you do not need to validate all the cells at once the feature's that we'd be focusing on will help. It's called eco-rendering and the change is that only the changed cell is being rendered after validation.
I think that it would be a good idea to add `validateCell` method after adding eco-renders to allow us to trigger the process for a single cell.

**Comment by Whythawye** [comment]
@AMBudnik Thanks for your reply! That would help with some of the performance issues we're running into, but unfortunately not with the core issue we're having. One of the core features of our app is to validate an entire csv file, and we are using handsontable to accomplish this. So all of the cells will need to be validated. With the eco-rendering, do you think it would be feasible to somehow run multiple validations in parallel instead of in sequence, perhaps?

**Comment by AMBudnik** [comment]
It hasn't been started yet so it's hard to tell.

**Comment by Whythawye** [comment]
Is there a timescale on when it might be started/finished?

**Comment by AMBudnik** [comment]
We are planning to start eco-renderers until the end of the year. But as it's something new I do not know how long it will take to get a working and bugless version.

**Comment by Whythawye** [comment]
Alright, thanks! It looks like we'll need to roll our own validation
system, then.

**Comment by AMBudnik** [comment]
I got `0.316s` with the v0.23.0 of HOT and `0.269s` with v0.35.0 (latest). But when I use a `1000 x 1000` array the time is `25.871s`.

Tested on Windows 10/ Chrome 63

**Comment by mkhoa1412** [comment]
Any update here ?

**Comment by AMBudnik** [comment]
Sorry, @mkhoa1412 do updates yet. All changes and idea are posted within the post.

**Comment by NoNameProvided** [comment]
This should be on a higher priority than normal. Handsontable is advertised as an excel like a solution for the web, but any table with more than 500cells with even a little-attached handling/validation logic becomes totally unusable due to performance reasons (even on high-end PCs). This is __a serious problem__ and I would expect that this can be resolved in 2.5 years.

**Comment by AMBudnik** [comment]
The performance issues are always tricky. It is not that we did not want to take our hands on it. Each time we change anything in the code we check if that change does not affect the performance of CRUD operations. Usually, it does, it gives or takes a fraction of milliseconds for the operation to finish. This in the end for a larger table can cause bigger changes as the result of a change multiplies by the number of cells.

Currently, to be honest, we are not working on any project that can boost the overall performance. Here is a list of tasks to be done in October https://github.com/handsontable/handsontable/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22October+2018%22
The topic that surely wil boost the perofmance of scrolling, editing and validation is the single cell renderer or what I like to call it eco-rendering. This is a project that will allow the developer to point out the cells to get rendered. As we know less calls to the renderer equals greater performance. 
However, this project is extremly complicated. Each time we would like to sit down and start to code we find `regression` type of bugs and `high priority` fixes that need to go first. Additionally the effective work of the `updateSettings` method, that you possibly know, is a must-have. Unfortunately now it is not fully operative. It needs to be fixed and possibly refecatored before we can move on with the performance booster - the eco-rendering.

I highly encourage to track [this topic on our Roadmap](https://trello.com/c/m9KTygSV/80-rewrite-renderers-to-improve-performance-2152) for fresh updates.

**Comment by AMBudnik** [comment]
I'm updating the demo to `8.0.0-beta1` http://jsfiddle.net/ko2ea1f7/

**Comment by allanbowe** [comment]
We are also affected by this issue.  Our workaround is to validate row by row.

**Comment by budnix** [comment]
I’ve tested most of the described cases across multiple Handsontable versions, including the latest one.
From our observations, the performance is now more than 2x faster compared to v9 (https://jsfiddle.net/budnix/qszcx2fL/3/)

Closing this issue for now. If you’re still experiencing any problems or notice regressions, don’t hesitate to open a new issue.
